### DISCUSSION TITLE
Discussion on the necessity of #cscope# functionality

### DISCUSSION THREAD

**Discussion by daishengdong** [neovim/neovim#25784] [General]
I am a beginner user of Neovim, and I would like to raise a concern regarding the removal of cscope support in Neovim since version 0.9.0. There may be some aspects that I have not considered, so I would like to discuss it here.

In my experience, the Language Server Protocol (LSP) used with Neovim, specifically clangd, does not fully replace the functionality of cscope. Here are a couple of reasons:

- In Neovim, when using Telescope to search for symbols in the workspace, if the project is large and the symbol is defined in a file that is not currently opened, the symbol is unlikely to appear in the candidate list. This poses a significant problem for symbol searching and navigation. To solve this problem, I usually rely on ctags for navigation (fortunately, Neovim still supports ctags).

- If a code block is wrapped in a configuration macro that is disabled, the symbols defined within that code block will not be recognized as normal symbols by the language server. This means that it becomes impossible to perform reference or definition jumps between these symbols. This behavior of the language server can be frustrating in daily development, especially if the project contains numerous code segments wrapped in configuration macros. However, this issue may not be applicable to other programming languages.

Apart from the functional issues with the language server, there are also deployment concerns related to its proper functioning. These concerns include:

- For example, clangd requires a compile_commands.json file, which is generated in makefile projects using other tools like bear or compiledb. However, this relies on the condition that the project can be successfully built on the development machine. This condition may not be easily met, such as when developing Linux kernel modules on macOS or Windows.

- Alternatively, deployment of tools like compiledb or clangd itself may encounter difficulties due to issues like network connectivity, incompatible Python versions, or old glibc versions on the development machine.

In comparison, cscope, as an "offline" symbol reference and definition navigation tool based on plain text parsing, offers certain advantages in terms of functionality and deployment. It can serve as a useful complement to LSP. I am curious about the community's considerations regarding the removal of cscope support in Neovim, as retaining this functionality does not seem to be a significant burden and can be highly beneficial in certain scenarios.

**Comment by zeertzjq** [comment]
Try https://github.com/dhananjaylatkar/cscope_maps.nvim

**Reply by daishengdong** [reply]
Thank you, I will give it a try and see.

**Comment by bfredl** [comment]
It seems the doc update to `:help cscope` mention cscope-plugins (instead of LSP) as the direct replacement was not backported to `release-0.9` . 

For reference this is the updated version https://neovim.io/doc/user/vim_diff.html#cscope

**Reply by daishengdong** [reply]
Ah, it seems that cscope_maps.nvim is an "officially recommended" plugin-based cscope solution. After using it for an afternoon, I can say that this plugin completely meets my daily development needs. Thank you all for your excellent work!
